### DISCUSSION TITLE
Attach a X-axis/bar of Pseudo-time value/cell cluster color bar on the top of the plotTrajectoryHeatmap

### DISCUSSION THREAD

**Discussion by willey2020** [GreenleafLab/ArchR#580] [Feature Requests]
Hello! Thanks again for this great package!

**Describe the problem that your feature request would address.**

The plotTrajectoryHeatmap function is very cool.
The y axis of the current plots(TF motif,Gene score, RNA) is very clear. However, it is better to get an accurate sense about which time point, a feature(motif,gene,RNA) is showing up/down, maybe through a more detailed x-axis or color bar, for example, the pseudo-time x axis or the pesudotime/cell cluster color bars, on the top of the trajectory heatmap. 
 
**Describe the solution you'd like**
Is there any way to attach a x-axis that shows pesudotime value according to the heatmap,like below or on the top of the heatmap? Then, we can get a sense of exactly when a TF/RNA/gene score is upregulated/downregulated.

**Describe alternatives you've considered**
If there is another way, that shows the matched color the pesudotime value (dark blue, purple to yellow, showed in manual 16.1.1) in a horizontal color bar plot, and matching the trajectory heatmap x-axis, then it will be much more clear.  

Or if there is a way to shows the cell cluster color bar together matching the x-axis, it will be great!

For these two alternative solutions, I attached two reference plots, examples from single-cell RNA-seq study. Just for examples only. For the second example, their cell type/stage colors showing on the top of the heatmap are a similar analogy to cell cluster color. 

**Additional context**
After searching in the issues, not sure if a similar feature request was posted before. 
Thank you so much and again this is a great and amazing package!

![Screenshot 2021-02-14 085109](https://user-images.githubusercontent.com/72377061/107883144-fa1e1e80-6ea1-11eb-9cef-f6311093dc87.png)
![Screenshot 2021-02-14 084132](https://user-images.githubusercontent.com/72377061/107882881-a52dd880-6ea0-11eb-84a7-20a26e3aac3c.png)

**Comment by rcorces** [comment]
I dont think this is currently possible but seems reasonable. Feature requests have piled up so I wouldnt expect this to be addressed soon though.

**Comment by willey2020** [comment]
Thank you!! Totally understand! Thanks again for this great package!

**Comment by hjc0818** [comment]
Hi there, I am using the package and found that this feature is important. While I found some code in their Nature Genetic paper can do this. Please see below

https://github.com/winstonbecker/scCRC_continuum/blob/main/scATAC_stromal.R

line 1028 - 1039

` trajGIM <- getTrajectory(ArchRProj = proj_stromal, name = "control_caf_trajectory", useMatrix = "GeneIntegrationMatrix", log2Norm = TRUE)
p3 <- plotTrajectoryHeatmap(trajGIM,  pal = paletteContinuous(set = "blueYellow"), labelTop = 20, varCutOff = 0.6)
plotPDF(p3, name = "Plot-control_caf_trajectory-GIM-Heatmap.pdf", ArchRProj = proj_stromal, addDOC = FALSE, width = 6, height = 10)
test <- getCellColData(proj_stromal)[!is.na(getCellColData(proj_stromal)$control_caf_trajectory),c("CellType", "control_caf_trajectory")]
 as.matrix(test[order(test$control_caf_trajectory),"CellType"])
 new_df <- data.frame(cbind(test[order(test$control_caf_trajectory),"CellType"], c(1:3043), rep(1,3043)))
 new_df$X2 <- as.numeric(as.character(new_df$X2))
 p <- ggplot(new_df, aes(X2, X3)) + geom_tile(aes(fill = X1)) + 
        scale_fill_manual(values= c("#272E6A", "#0C727C", "#6E4B9E"))+theme_ArchR()+theme(axis.text.x = element_text(angle = 90))
 plotPDF(p, name = paste0("caf_control_trajectory_hm.pdf"), width = 5, height = 5, ArchRProj = proj_stromal, addDOC = FALSE)
`

<img width="1577" alt="image" src="https://user-images.githubusercontent.com/45670154/195142880-d0514b06-00da-44c7-baf8-5b520d311d54.png">

Hope it helps, and I hope someone can incorporate this code into the plotTrajectoryHeatmap function, since right now its just giving out a heatmap without any label on top. Will be very frustrating for non-hardcore users.

Sincerely

Han.

**Reply by rcorces** [reply]
@hjc0818 - thanks for the reminder. I had half-way coded this and lost track of it.
I've incorporated this into the `dev_trajectoryHeatmap` branch. You can install that branch like so:
```
devtools::install_github("GreenleafLab/ArchR", ref="dev_trajectoryHeatmap", repos = BiocManager::repositories())
#to unload a package and reload
detach("package:ArchR", unload=TRUE)
library(ArchR)
```

After installing, you will need to re-run `getTrajectory()` paying special attention to the new `trajectoryLabel` parameter:
https://github.com/GreenleafLab/ArchR/blob/586c8a908df3404a3852a171439048fd35833231/R/Trajectory.R#L290-L292

Then you will run `plotTrajectoryHeatmap()` and include the new `colorColumns` and `columnPal` parameters:

https://github.com/GreenleafLab/ArchR/blob/586c8a908df3404a3852a171439048fd35833231/R/Trajectory.R#L470-L475

Please let me know if this works for you. If so, I will merge these changes into `dev` for future incorporation into `release_1.0.3`

Example below:
```
trajectory <- c("Progenitor", "GMP", "Mono")

projHeme5 <- addTrajectory(
  ArchRProj = projHeme5, 
  name = "MyeloidU", 
  groupBy = "Clusters2",
  trajectory = trajectory, 
  embedding = "UMAP", 
  force = TRUE
)

trajMM  <- getTrajectory(ArchRProj = projHeme5, name = "MyeloidU", useMatrix = "MotifMatrix", log2Norm = FALSE, trajectoryLabel = "Clusters2")

p1 <- plotTrajectoryHeatmap(trajMM, pal = paletteContinuous(set = "solarExtra"), colorColumns = TRUE, columnPal = paletteDiscrete(values = unique(colData(trajMM)$label)))

plotPDF(p1, name = "Plot-MyeloidU-Traj-Heatmaps.pdf", ArchRProj = projHeme5, addDOC = FALSE, width = 6, height = 8)
```
![image](https://user-images.githubusercontent.com/10343046/195213655-31dc382f-1874-42b3-af94-24fddf81f78b.png)

**Reply by rcorces** [reply]
@hjc0818 - could you test this out please?

**Comment by taminastelzer** [comment]
Hi, first of all thanks for the great tool. I really enjoy working with ArchR. I am currently doing a lot of Pseudotime Trajectory analysis. For this, I would like to add a X-axis/bar of cell cluster on the top of the plotTrajectoryHeatmap. While searching for this possibility, I came across this issue. However, unfortunately I am not able to install the `dev_trajectoryHeatmap` branch. I receive following Error message: 

`Downloading GitHub repo GreenleafLab/ArchR@dev_trajectoryHeatmap
Error in utils::download.file(url, path, method = method, quiet = quiet,  : 
  cannot open URL 'https://api.github.com/repos/GreenleafLab/ArchR/tarball/dev_trajectoryHeatmap'`

Is this branch not present anymore? I would highly appreciate your help!

Thanks a lot!

**Reply by rcorces** [reply]
this was merged into the `dev` branch.
